### PR TITLE
[jit] fix aten::grad to return optional list

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29577 [jit] fix aten::grad to return optional list**

Summary:

`torch.autograd.grad` can return none is one of the input is not in the
autograd graph or not requires_grad, this fix it so that it return a
list of optional tensor instead of list of tensor.

This might have BC issue unfortunately, but I think it's rare both
internal and external (only training use it, and most of the training
use backward, instead of autograd.grad), so whitelist it.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18491642](https://our.internmc.facebook.com/intern/diff/D18491642)

### CODE REVIEW

**Comment by eellison** [comment]
List[Tensor] doesn't subtype List[Optional[Tensor]]

**Comment by wanchaol** [comment]
> List[Tensor] doesn't subtype List[Optional[Tensor]]

@eellison hmmm ok, but List[Tensor] can be passed into a function that is accepting List[Optional[Tensor]] since we try to unify the inner types, which should resolve the BC concern.

**Comment by eellison** [comment]
> > List[Tensor] doesn't subtype List[Optional[Tensor]]
> 
> @eellison hmmm ok, but List[Tensor] can be passed into a function that is accepting List[Optional[Tensor]] since we try to unify the inner types, which should resolve the BC concern.

I don't think List[Tensor] can't be passed into a function that accepts List[Optional[Tensor]]

**Comment by wanchaol** [comment]
> > > List[Tensor] doesn't subtype List[Optional[Tensor]]
> > 
> > 
> > @eellison hmmm ok, but List[Tensor] can be passed into a function that is accepting List[Optional[Tensor]] since we try to unify the inner types, which should resolve the BC concern.
> 
> I don't think List[Tensor] can't be passed into a function that accepts List[Optional[Tensor]]

it can I think, the following code works:

```python
@torch.jit.script 
def return_tensor_list(): 
	# type: () -> List[Tensor] 
    return [torch.randn(3, 2), torch.randn(3, 2)] 

@torch.jit.script 
def test_use_opt_list(t_list): 
    # type: (List[Optional[Tensor]]) -> Optional[Tensor] 
    return t_list[0]

test_use_opt_list(return_tensor_list()) 

Out[5]: 
tensor([[ 0.5223, -0.9674],
        [ 0.2795, -0.4572],
        [-0.0400, -1.8270]])
```

**Comment by eellison** [comment]
> > > > List[Tensor] doesn't subtype List[Optional[Tensor]]
> > > 
> > > 
> > > @eellison hmmm ok, but List[Tensor] can be passed into a function that is accepting List[Optional[Tensor]] since we try to unify the inner types, which should resolve the BC concern.
> > 
> > 
> > I don't think List[Tensor] can't be passed into a function that accepts List[Optional[Tensor]]
> 
> it can I think, the following code works:
> 
> ```python
> @torch.jit.script 
> def return_tensor_list(): 
> 	# type: () -> List[Tensor] 
>     return [torch.randn(3, 2), torch.randn(3, 2)] 
> 
> @torch.jit.script 
> def test_use_opt_list(t_list): 
>     # type: (List[Optional[Tensor]]) -> Optional[Tensor] 
>     return t_list[0]
> 
> test_use_opt_list(return_tensor_list()) 
> 
> Out[5]: 
> tensor([[ 0.5223, -0.9674],
>         [ 0.2795, -0.4572],
>         [-0.0400, -1.8270]])
> ```

That's at the pybind layer so it's different because inputs from python don't have a well-formed type yet. Try:

```
@torch.jit.script 
def return_tensor_list(): 
	# type: () -> List[Tensor] 
    return [torch.randn(3, 2), torch.randn(3, 2)] 

@torch.jit.script 
def test_use_opt_list(t_list): 
    # type: (List[Optional[Tensor]]) -> Optional[Tensor] 
    return t_list[0]

@torch.jit.script
def will_fail():
    test_use_opt_list(return_tensor_list()) 
```

**Review by zdevito — APPROVED** [review]
We should land this, but we need to be careful because it does break BC. Why is it ok to break BC?

1. It is a bugfix, so the old behavior was not correct when gradients were None.
2. It is unlikely to be used in any TorchScript inference models.
3. grad was introduced relatively recently, so we do not expect a ton of serialized use of it.

Before landing, we should check that we are not breaking known predictor models.
